### PR TITLE
fix: Add retry logic for GET operations when no peers available

### LINKED ISSUE #1858: GET operation fails immediately when no peers available instead of retrying
## Description

When a peer has no other peers available to forward a GET request to, the operation fails immediately with "reached max retries" instead of actually retrying. This prevents content retrieval in sparse network conditions.

## Confirmed Bug

The GET operation code has a logic flaw: when `k_closest_potentially_caching()` returns an empty list (no available peers), it immediately returns failure WITHOUT:
1. Actually incrementing the retry counter
2. Implementing any delay mechanism  
3. Attempting the configured MAX_RETRIES (10)

### User's peer logs:
```
2025-09-26T02:48:53.726028Z  INFO freenet::operations::connect: Immediately requesting more peer connections from gateway

[3.5 minutes pass - user attempts to join River room]

2025-09-26T02:52:23.003037Z  INFO freenet::operations::get: Seek contract, tx: 01K61YZ7JT1PP4WGM8N4ZRDWG2, key: 9L1N9DyVwofcib7PpQqjsmdkcEWLBf8PUVGMLx9LDW1H, target: v6MWKgqHiBMNcGtG

2025-09-26T02:52:23.195452Z  WARN freenet::operations::get: Neither contract or contract value for contract found at peer v6MWKgqHiBMNcGtG, retrying with other peers

2025-09-26T02:52:23.195511Z ERROR freenet::operations::get: Failed getting a value for contract 9L1N9DyVwofcib7PpQqjsmdkcEWLBf8PUVGMLx9LDW1H, reached max retries
```
**Note**: Entire operation failed in ~192ms, clearly not attempting 10 retries.

### Gateway logs (same transaction):
```
2025-09-26T02:52:23.092836Z  WARN freenet::operations::get: No other peers found while trying to get the contract, tx: 01K61YZ7JT1PP4WGM8N4ZRDWG2, key: 9L1N9DyVwofcib7PpQqjsmdkcEWLBf8PUVGMLx9LDW1H, this_peer: v6MWKgqHiBMNcGtG
    at crates/core/src/operations/get.rs:1135
```

## Root Cause

In [`get.rs` lines 695-744](https://github.com/freenet/freenet-core/blob/main/crates/core/src/operations/get.rs#L695-L744), when `new_candidates.is_empty()`:
- If there's a requester peer: returns failure immediately
- If original requester: logs "reached max retries" and fails
- Never actually increments retry counter or delays

## Why the Gateway Had No Peers (Hypothesis)

Several possible explanations:
1. **Version incompatibility**: The v0.1.27 release had just happened, possibly leaving the gateway isolated from older versions
2. **Network bootstrap issue**: The gateway may have lost connections and hadn't re-established them
3. **Actual sparse network**: Simply very few peers online
4. **Connection failures**: Network issues preventing peer connections

We cannot determine the exact cause from the logs alone.

## Impact

This bug affects any scenario where peers temporarily have no connections:
- Network bootstrap/startup
- Version transitions (if versions are incompatible)
- Network partitions
- Connection losses
- Small/test networks

## Proposed Solution

The GET operation needs proper retry logic:
1. **Always increment retry counter**, even when no peers available
2. **Implement exponential backoff delays** (100ms, 200ms, 400ms, etc.)
3. **Actually attempt MAX_RETRIES times** before failing
4. **Consider scheduling retries** rather than blocking

This would allow time for:
- Peer connections to establish
- Network conditions to improve
- Bootstrap processes to complete

## Additional Issues

1. **Misleading error message**: "reached max retries" when no retries were attempted
2. **No grace period**: Fails immediately even though peer requested more connections
3. **Test gap**: Tests use pre-populated networks, missing this edge case

## Related Code References

- Bug location: [`get.rs:695-744`](https://github.com/freenet/freenet-core/blob/main/crates/core/src/operations/get.rs#L695-L744)
- MAX_RETRIES constant: [`get.rs:30`](https://github.com/freenet/freenet-core/blob/main/crates/core/src/operations/get.rs#L30)
- Gateway's "no peers" warning: [`get.rs:1135`](https://github.com/freenet/freenet-core/blob/main/crates/core/src/operations/get.rs#L1135)

[AI-assisted debugging and comment]

### PR DESCRIPTION
**Pull request by sanity**
## Summary
This PR addresses issue #1858 where GET operations fail immediately when no peers are available instead of utilizing the retry budget (MAX_RETRIES=10).

[AI-assisted debugging and comment]

## Problem
When `k_closest_potentially_caching()` returns an empty list (no peers available), the GET operation immediately falls into the failure branch without checking or using the retry counter. This causes operations to fail in ~192ms instead of retrying with exponential backoff as intended.

## Solution
- Added retry counter check before failing when no peers are available
- Preserves operation state with incremented retry counter for potential retries
- Added regression test to verify retry behavior

## Implementation Notes
**Important:** This is a partial fix that prevents immediate failure but doesn't automatically schedule retries. The current implementation:
1. Checks if retries < MAX_RETRIES before giving up
2. Preserves the operation state with incremented retry counter
3. Logs the retry attempt for debugging

Full automatic retry scheduling would require deeper integration with the state machine architecture and should be reviewed by the team for the best approach.

## Test Plan
- [x] Added regression test `test_get_retry_with_no_peers` that verifies GET operations don't fail immediately
- [ ] Manual testing with isolated nodes
- [ ] Verify existing tests still pass

## Related Issues
Fixes #1858

## Additional Context
During testing, I discovered a separate issue where the client isn't notified when `request_get` returns an error (line 837 in client_events/mod.rs). This causes the client to hang waiting for a response. This should be addressed separately.

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: crates/core/tests/get_retry_regression.rs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        // This test verifies the fix prevents immediate failure, but operation will timeout
+        // waiting for response since client notification issue exists
+        let get_result = timeout(Duration::from_secs(30), client.recv()).await;
+        let get_elapsed = get_start.elapsed();
+
+        println!("GET operation completed in {:?}", get_elapsed);
+
+        // The operation will timeout because of client notification issue
+        // But we can verify it didn't fail immediately (< 1 second)
+        assert!(
```
Comment (iduartgomez) [inline crates/core/tests/get_retry_regression.rs:149]: idk how good this test is, we should use exp backoff for the retries to ensure we reach an expected Duration at the very least

File: crates/core/src/operations/get.rs [hunk 2]
```diff
@@ -719,35 +719,71 @@ impl Operation for GetOp {
...
-                                    tracing::error!(
-                                        tx = %id,
-                                        "Failed getting a value for contract {}, reached max retries",
-                                        key
-                                    );
-                                    return_msg = None;
-                                    new_state = None;
+                                    // No peers available right now
+                                    // Check if we should retry with exponential backoff
+                                    const MAX_RETRIES: usize = 10;
```
Comment (iduartgomez) [inline crates/core/src/operations/get.rs:725]: we probably should use our backoff utils for this

**Comment by sanity** [comment]
[Codex] Thanks for jumping on this so quickly! I pulled the branch and I think we still have a couple of big gaps:

1. When `k_closest_potentially_caching` returns an empty list we now set `new_state = Some(GetState::AwaitingResponse { retries: retries + 1, ... })` and return `OperationResult { return_msg: None, state: Some(..) }`. That pushes the op back into `OpManager`, but nothing ever wakes it up again—there is no timer, backoff task, or `missing_candidate_peers` notification that will re-run `SeekNode`. The next thing that happens is the TTL expires and `garbage_cleanup_task` fires `TransactionTimedOut`, which finally produces the error we see in the logs. So even if the node discovers new peers a few seconds later, the GET never retries. Could we either schedule an explicit retry (e.g. via `tokio::spawn` + sleep/backoff) or hook into the connection event that already exists for "missing candidate" peers?

2. The regression test passes because it now waits ~60s (the operation TTL) before failing, but that’s effectively the same as raising the timeout—it doesn’t exercise actual retry logic. We should either prove that a second GET attempt is emitted when a new peer appears, or tighten the test to mock the retry path.

3. Minor nit: there’s a new `const MAX_RETRIES` inside the branch even though we already have the module-level constant.

Given those points, I think the fix would still leave #1858 in place: GET would just hang for the full TTL rather than retrying when new peers appear. Happy to pair on wiring this into the existing connection notifications if that helps.

**Comment by iduartgomez** [comment]
> [Codex] Thanks for jumping on this so quickly! I pulled the branch and I think we still have a couple of big gaps:
> 
>     1. When `k_closest_potentially_caching` returns an empty list we now set `new_state = Some(GetState::AwaitingResponse { retries: retries + 1, ... })` and return `OperationResult { return_msg: None, state: Some(..) }`. That pushes the op back into `OpManager`, but nothing ever wakes it up again—there is no timer, backoff task, or `missing_candidate_peers` notification that will re-run `SeekNode`. The next thing that happens is the TTL expires and `garbage_cleanup_task` fires `TransactionTimedOut`, which finally produces the error we see in the logs. So even if the node discovers new peers a few seconds later, the GET never retries. Could we either schedule an explicit retry (e.g. via `tokio::spawn` + sleep/backoff) or hook into the connection event that already exists for "missing candidate" peers?
> 
>     2. The regression test passes because it now waits ~60s (the operation TTL) before failing, but that’s effectively the same as raising the timeout—it doesn’t exercise actual retry logic. We should either prove that a second GET attempt is emitted when a new peer appears, or tighten the test to mock the retry path.
> 
>     3. Minor nit: there’s a new `const MAX_RETRIES` inside the branch even though we already have the module-level constant.
> 
> 
> Given those points, I think the fix would still leave #1858 in place: GET would just hang for the full TTL rather than retrying when new peers appear. Happy to pair on wiring this into the existing connection notifications if that helps.

A short-term solution can be to exercise the retry inside the op instead of waiting for it to be scheduled again.

As for the test, seems the right test would be to boot a 2nd peer after X secs and assert that the 1st peer was indeed able to get the contract with a single operation.

In all honesty, all this could be handled app side (and probably should), as long as we return meaningful messages to the app/client. If you are disconnected to the network, yes, your network operations will fail, and following the principle of least surprise would be informing the consumer about it. Is one thing trying to be smart around transient issues, but this does not seem to be that case.

Idk why we thought we should otherwise would be better behavior, I personally would dislike that as app developer.

**Comment by sanity** [comment]
@iduartgomez Ok, so should we abandon this PR and instead just ensure that failures are properly reported to the client?

**Comment by sanity** [comment]
Closing this PR based on team discussion. 

@iduartgomez is correct that retry logic should be handled at the application level rather than in the network layer. The network layer should focus on:

1. **Properly reporting errors to clients** (the real bug - clients currently hang when operations fail)
2. **Providing meaningful error messages** ("No peers available", "Network disconnected", etc.)
3. **Allowing applications to make retry decisions** based on their own context

The current approach in this PR:
- Doesn't actually schedule retries (no timer/wakeup mechanism)
- Just delays failure from 192ms to 60s TTL timeout
- Overcomplicates the operation state machine
- Doesn't fix the root cause (client not notified of errors)

Opening a new PR to fix the actual bug: ensuring operation errors are properly reported to WebSocket clients with meaningful error messages.

Related: #1858

[AI-assisted debugging and comment]
